ursa/graph/utils.py

```diff
@@ -44,3 +47,64 @@ def _apply_append(collection, values):
             collection.update(val)
 
         return collection
+
+
+@ray.remote(num_return_vals=0)
+def write_vertex(vertex, graph_id, key):
+    dest_directory = graph_id + "/" + key + "/"
+
+    # @TODO(kunalgosar): Accessing _transaction_id here is not recommended
+    dest_file = dest_directory + str(vertex._transaction_id) + ".dat"
+    if not os.path.exists(dest_directory):
+        os.makedirs(dest_directory)
+
+    serialization_context = ray.worker.global_worker.serialization_context
+
+    oids = [vertex.vertex_data]
+    oids.extend(vertex.local_edges.edges)
+    local_edges_buffer = pa.serialize(vertex.local_edges.buf,
+                                      serialization_context).to_buffer()
+
+    oids.append(local_edges_buffer)
+    num_local_edge_lists = len(vertex.local_edges.edges) + 1
+
+    foreign_edges = list(vertex.foreign_edges.keys())
+    oids.extend([vertex.foreign_edges[k] for k in foreign_edges])
+    oids = [pa.plasma.ObjectID(oid.id()) for oid in oids
+            if type(oid) is ray.local_scheduler.ObjectID]
+
+    buffers = ray.worker.global_worker.plasma_client.get_buffers(oids)
+    data = {"vertex_data": buffers[0],
+            "local_edges": buffers[1:1 + num_local_edge_lists],
+            "foreign_edges": foreign_edges,
+            "foreign_edge_values": buffers[1 + num_local_edge_lists:],
+            "transaction_id": vertex._transaction_id}
+
+    serialized = pa.serialize(data, serialization_context).to_buffer()
+    with pa.OSFile(dest_file, 'wb') as f:
+        f.write(serialized)
+
+
+@ray.remote
+def read_vertex(file):
+    mmap = pa.memory_map(file)
+    buf = mmap.read()
+
+    serialization_context = ray.worker.global_worker.serialization_context
+    data = pa.deserialize(buf, serialization_context)
+
+    vertex_data = pa.deserialize(data["vertex_data"], serialization_context)
+    local_edges = np.array([pa.deserialize(x, serialization_context)
+                            for x in data["local_edges"]]).flatten()
+
+    foreign_edge_values = [pa.deserialize(x, serialization_context)
+                           for x in data["foreign_edge_values"]]
+
+    foreign_edges = dict(zip(data["foreign_edges"], foreign_edge_values))
+
+    new_vertex = ursa.graph.vertex._Vertex(vertex_data,
```

Comment:
Is this (`ursa.graph.vertex._Vertex`) currently the only way to get the` _Vertex` class?

Reply:
I believe that there should be a better way of doing this, but nothing seemed to work. I think there is some discrepancy between the environment of the file and the environment that the function is run in when called from outside.

Devin had also taken a look at this today and did not find anything.

Reply:
Okay, thanks for the info. We may come back to this in the future and try to clean it up.